acra-censor/common/logging_logic.go

```diff
@@ -0,0 +1,285 @@
+					handler.logger.WithError(ErrComplexSerializationError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't create QueryWriter instance")
+				}
+				return
+
+			default:
+				//do nothing. This means that channel has no data to read yet
+			}
+		}
+	}()
+	return handler, nil
+}
+
+// SetSerializationTimeout sets timeout of dumping captured queries to the file.
+func (queryWriter *QueryWriter) SetSerializationTimeout(timeout time.Duration) {
+	queryWriter.serializationTimeout = timeout
+}
+
+// GetSerializationTimeout gets timeout of dumping captured queries to the file.
+func (queryWriter *QueryWriter) GetSerializationTimeout() time.Duration {
+	return queryWriter.serializationTimeout
+}
+
+// DumpAllQueriesToFile writes stored queries into file.
+func (queryWriter *QueryWriter) DumpAllQueriesToFile() error {
+	// open or create file, NO APPEND
+	f, err := os.OpenFile(queryWriter.filePath, os.O_TRUNC|os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		queryWriter.logger.WithError(ErrCantOpenFileError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't dump queries to file")
+		return err
+	}
+	// write all queries
+	return queryWriter.appendQueries(queryWriter.Queries, f)
+}
+
+// Release dumps all Captured queries to file and resets Captured queries list.
+func (queryWriter *QueryWriter) Release() {
+	queryWriter.signalBackgroundExit <- true
+	queryWriter.reset()
+}
+
+func (queryWriter *QueryWriter) reset() {
+	queryWriter.Queries = nil
+	queryWriter.BufferedQueries = nil
+}
+
+func (queryWriter *QueryWriter) readAllQueriesFromFile() error {
+	q, err := readQueries(queryWriter.filePath)
+	if err != nil {
+		queryWriter.logger.WithError(ErrCantReadQueriesFromFileError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't read queries from file")
+		return err
+	}
+	// read existing queries from file
+	queryWriter.Queries = q
+	return nil
+}
+
+func readQueries(filePath string) ([]*QueryInfo, error) {
+	bufferBytes, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+	var queries []*QueryInfo
+	if len(bufferBytes) != 0 {
+		for _, line := range bytes.Split(bufferBytes, []byte{'\n'}) {
+			if len(line) == 0 {
+				continue
+			}
+			var oneQuery QueryInfo
+			if err = json.Unmarshal(line, &oneQuery); err != nil {
+				return nil, err
+			}
+			queries = append(queries, &oneQuery)
+		}
+	}
+	return queries, nil
+}
+
+func (queryWriter *QueryWriter) dumpBufferedQueriesToFile(openedFile *os.File) error {
+	// nothing to dump
+	if len(queryWriter.BufferedQueries) == 0 {
+		return nil
+	}
+	err := queryWriter.appendQueries(queryWriter.BufferedQueries, openedFile)
+	if err != nil {
+		return err
+	}
+	// clean buffered queries only after successful write
+	queryWriter.BufferedQueries = nil
+	return nil
+}
+
+func (queryWriter *QueryWriter) appendQueries(queries []*QueryInfo, openedFile *os.File) error {
```

Comment:
because we don't use any os.File specific API then we can take io.Writer as param.

Reply:
We use `openedFile.Write(lines)` in this function